crypto/jwks/component.go

```diff
@@ -0,0 +1,248 @@
+	"time"
+
+	"github.com/lestrrat-go/httprc"
+	"github.com/lestrrat-go/jwx/v2/jwk"
+
+	contribCrypto "github.com/dapr/components-contrib/crypto"
+	"github.com/dapr/kit/fswatcher"
+	"github.com/dapr/kit/logger"
+)
+
+const (
+	defaultRequestTimeout            = 30 * time.Second
+	metadataKeyJWKS                  = "jwks"
+	metadataKeyRequestTimeoutSeconds = "requestTimeoutSeconds"
+
+	// Minimum interval before refreshing the JWKS cache
+	minRefreshInterval = 10 * time.Minute
+)
+
+type jwksCrypto struct {
+	contribCrypto.LocalCryptoBaseComponent
+
+	requestTimeout time.Duration
+
+	jwks     jwk.Set
+	jwksLock sync.Mutex
+
+	logger logger.Logger
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+// NewJWKSCrypto returns a new crypto provider based a JWKS, either passed as metadata, or read from a file or HTTP(S) URL.
+// The key argument in methods is the ID of the key in the JWKS ("kid" property).
+func NewJWKSCrypto(logger logger.Logger) contribCrypto.SubtleCrypto {
+	ctx, cancel := context.WithCancel(context.Background())
+	k := &jwksCrypto{
+		logger: logger,
+		ctx:    ctx,
+		cancel: cancel,
+	}
+	k.RetrieveKeyFn = k.retrieveKeyFromSecretFn
+	return k
+}
+
+// Init the crypto provider.
+func (k *jwksCrypto) Init(ctx context.Context, metadata contribCrypto.Metadata) error {
+	if len(metadata.Properties) == 0 {
+		return errors.New("empty metadata properties")
+	}
+
+	if metadata.Properties[metadataKeyRequestTimeoutSeconds] != "" {
+		timeoutSec, _ := strconv.Atoi(metadata.Properties[metadataKeyRequestTimeoutSeconds])
+		if timeoutSec > 0 {
+			k.requestTimeout = time.Duration(timeoutSec) * time.Second
+		}
+	}
+	if k.requestTimeout == 0 {
+		k.requestTimeout = defaultRequestTimeout
+	}
+
+	err := k.initJWKS(ctx, metadata.Properties[metadataKeyJWKS])
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Close implements the io.Closer interface to close the component
+func (k *jwksCrypto) Close() error {
+	if k.cancel != nil {
+		k.cancel()
+	}
+	return nil
+}
+
+// Features returns the features available in this crypto provider.
+func (k *jwksCrypto) Features() []contribCrypto.Feature {
+	return []contribCrypto.Feature{} // No Feature supported.
+}
+
+// Init the JWKS object from the metadata property
+func (k *jwksCrypto) initJWKS(ctx context.Context, md string) error {
+	if len(md) == 0 {
+		return errors.New("metadata property 'jwks' is required")
+	}
+
+	// If the value starts with "http://" or "https://", treat it as URL
+	if strings.HasPrefix(md, "http://") || strings.HasPrefix(md, "https://") {
+		return k.initJWKSFromURL(ctx, md)
+	}
+
+	// Check if the value is a valid path to a local file
+	stat, err := os.Stat(md)
+	if err == nil && stat != nil && !stat.IsDir() {
+		return k.initJWKSFromFile(ctx, md)
+	}
+
+	// Treat the value as the actual JWKS
+	// First, check if it's base64-encoded (remove trailing padding chars if present first)
+	mdJSON, err := base64.RawStdEncoding.DecodeString(strings.TrimRight(md, "="))
```

Comment:
is a trim right needed here?

Reply:
No, with base64 the `=` can only be present at the end. See: https://en.wikipedia.org/wiki/Base64